Flutter SDK [DOC-811] (#6088)
segmentio/segment-docs: src/connections/sources/catalog/libraries/mobile/flutter/index.md

---
title: Analytics-Flutter
hidden: false
---

> info ""
> The Analytics-Flutter library is currently in public beta and is governed by Segment's [First Access and Beta Preview Terms](https://www.twilio.com/en-us/legal/tos){:target="_blank"}. For more information, see the [Analytics-Flutter GitHub repository](https://github.com/segmentio/analytics_flutter){:target="_blank"}. 

> Analytics-Flutter enables you to seamlessly add Segment analytics to your Flutter1 app.

Reviewer: Stupid question: is it actually "Flutter1" or was that a typo? 😅

Author: oops sorry, that's a typo

### Supported platforms
Analytics-Flutter supports these platforms:

* Android
* iOS
* MacOS
* Web

Some destination plugins may not support all platform functionality. Refer to individual platform SDKs for more details.

## Getting started
To install Analytics-Flutter:

1. Manually add this package to your `pubspec.yaml` file: 

    ```yml
    dependencies: 
        flutter: 
            sdk: flutter 

        # Main package
        analytics: 
            git: 
                url: https://github.com/segmentio/analytics_flutter
                ref: main
                path: packages/core
    ```
2. *(Optional)* Add any plugin that you need.

>     ```yml
>       # Plugins
>     analytics_plugin_firebase:
>         git: 
>             url: https://github.com/segmentio/analytics_flutter
>             ref: main
>             path: packages/plugins/plugin_firebase
>     ```

Reviewer 2: Also this:

```suggestion
    ```bash
    flutter pub add segment_analytics_plugin_firebase
    ```
```

3. Import the library in your Dart code.

    ```dart

>     import 'package:analytics/client.dart';

Reviewer 2: ```suggestion
    import 'package:segment_analytics/client.dart';
```

  custom: {
    "plan": "premium"
  }
);
```
{% endcodeexampletab %}
{% endcodeexample %}

### Group

The [Group](/docs/connections/spec/group/) method is how you associate an individual user with a group — whether it's a company, organization, account, project, team. This includes a unique group ID and any optional group traits you know about them like the company name, industry, the number of employees. The traits option can include any information you might want to tie to the group, but when using any of the [reserved group traits](/docs/connections/spec/group/#traits), you should make sure to only use them for their intended meaning. All reserved traits are strongly typed by the ```GroupTraits``` class. When you use traits not listsed as a reserved user trait, these go under the ```custom``` property.

{% codeexample %}
{% codeexampletab Method signature %}
```dart
Future group(String groupId, {GroupTraits? groupTraits});
```
{% endcodeexampletab %}
{% codeexampletab Example use %}
```dart
analytics.group("some-company", groupTraits: GroupTraits(
  name: 'Segment',
  custom: {
    "region": "UK"
  }
);
```
{% endcodeexampletab %}
{% endcodeexample %}

## Utility methods

> The Analytics-Flutter utility methods help you work with plugins fro the analytics timeline. They include:

Reviewer: Small type: "fro" -> "from"

```dart
analytics.reset();
```
{% endcodeexampletab %}
{% endcodeexample %}

### Flush
By default, the analytics send to the API after 30 seconds or when 20 items accumulate, whichever happens first, and whenever the app resumes if the user has closed the app with some events unsent. You can modify these values by the `flushAt` and `flushInterval` config options. You can also trigger a flush event manually.

{% codeexample %}
{% codeexampletab Method signature %}
```dart
Future flush();
```
{% endcodeexampletab %}
{% codeexampletab Example use %}
```dart
analytics.flush();
```
{% endcodeexampletab %}
{% endcodeexample %}

### Advanced cleanup

In case you need to reinitialize the client, because you called `createClient` more than once for the same client in your application lifecycle, use this method _on the old client_ to clear any subscriptions and timers first.

```dart
var analytics = createClient(Configuration(writeKey));

analytics.cleanup();

analytics = createClient(Configuration(writeKey));
```

If you don't do this, the old client instance still exists and retains the timers, which makes all of your events fire twice.

## Automatic screen tracking

> To set up automatic screen tracking, you need to add the analytics navigator observer to your app's navigator observers. For example, if you're using the `MaterialApp` class, add the following:

Reviewer: I was confused by what Automatic Screen Tracking was until I read more in the Github repo - maybe we should include this context in these docs also:

> Sending a `screen()` event with each navigation action will get tiresome, so you may want to track navigation globally.

| `enrichment`   | Executes as the first level of event processing.                                                        |
| `destination`  | Executes as events begin to pass off to destinations.                                                   |
| `after`        | Executes after all event processing completes.  This can be used to perform cleanup operations. |
| `utility`      | Executes only with manual calls such as Logging.                                                    |

Plugins can have their own native code (for example, the iOS-only `analytics_plugin_idfa`) or wrap an underlying library (such as `analytics_plugin_firebase` which uses `firebase_core` and `firebase_analytics` under the hood).

### Destination plugins

Segment is included as a `DestinationPlugin` out of the box. You can add as many destination plugins as you like and upload events and data to them.

You can pass `autoAddSegmentDestination = false` in the options when setting up your client to prevent the `SegmentDestination` plugin from being added automatically.

### Adding Plugins

You can add a plugin at any time through the Add method.

```dart

> import 'package:analytics/client.dart';
> import 'package:analytics/event.dart';
> import 'package:analytics/state.dart';
> import 'package:analytics_plugin_advertising_id/plugin_advertising_id.dart';
> import 'package:analytics_plugin_idfa/plugin_idfa.dart';
> import 'package:analytics_plugin_firebase/plugin_firebase.dart'
>     show FirebaseDestination;

Reviewer 2: ```suggestion
import 'package:segment_analytics/client.dart';
import 'package:segment_analytics/event.dart';
import 'package:segment_analytics/state.dart';
import 'package:segment_analytics_plugin_advertising_id/plugin_advertising_id.dart';
import 'package:segment_analytics_plugin_idfa/plugin_idfa.dart';
import 'package:segment_analytics_plugin_firebase/plugin_firebase.dart'
    show FirebaseDestination;
```

const writeKey = 'SEGMENT_API_KEY';

class _MyAppState extends State<MyApp> {
  final analytics = createClient(Configuration(writeKey));

  @override
  void initState() {
    super.initState();
    initPlatformState();

    analytics
        .addPlugin(FirebaseDestination(DefaultFirebaseOptions.currentPlatform));
    analytics.addPlugin(PluginAdvertisingId());
    analytics.addPlugin(PluginIdfa());
  }
}
```

### Writing your own plugins

Plugins are implemented by extending one of the provided plugin classes. The available plugin classes are:

- `Plugin`
- `EventPlugin`
- `DestinationPlugin`
- `UtilityPlugin`
- `PlatformPlugin`

Any plugin must be an extension of one of these classes.

You can them customize the functionality by overriding different methods on the base class. For example, here is a `Logger` plugin:

```dart
import 'dart:convert';

> import 'package:analytics/analytics.dart';
> import 'package:analytics/event.dart';
> import 'package:analytics/plugin.dart';
> import 'package:analytics/logger.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/analytics.dart';
import 'package:segment_analytics/event.dart';
import 'package:segment_analytics/plugin.dart';
import 'package:segment_analytics/logger.dart';
```

| [AppsFlyer](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_appsflyer){:target="_blank"}    | `analytics_plugin_appsflyer`|
| [Firebase](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_firebase){:target="_blank"}      | `analytics_plugin_firebase`|
| [IDFA](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_idfa){:target="_blank"}     | `analytics_plugin_idfa` |
| [Android Advertising ID](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_advertising_id){:target="_blank"} | `analytics_plugin_advertising-id` |
  
  
## Controlling upload with flush policies

You can use `FlushPolicies` to more granurily control when events upload.

A flush policy defines the strategy for deciding when to flush. This can be on an interval, on a certain time of day, after receiving a certain number of events, or even after receiving a particular event. This gives you more flexibility on when to send events to Segment.

To make use of flush policies, you can set them in the configuration of the client:

```dart

> import 'package:analytics/flush_policies/count_flush_policy.dart';
> import 'package:analytics/flush_policies/timer_flush_policy.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/flush_policies/count_flush_policy.dart';
import 'package:segment_analytics/flush_policies/timer_flush_policy.dart';
```

```dart
if (isConnected) {
  analytics.addFlushPolicy(policiesIfNetworkIsUp);
} else {
  analytics.removeFlushPolicy(policiesIfNetworkIsUp)
}
```

### Creating your own flush policies

You can create a custom `FlushPolicy` for your application needs by implementing the  `FlushPolicy` interface. You can also extend the `FlushPolicyBase` class that already creates and handles the `shouldFlush` value reset.

A `FlushPolicy` only needs to implement 1 method:
- `onEvent(RawEvent event)`: Gets called on every event tracked by your client.

A `FlushPolicy` can optionally implement:
- `reset()`: Calls after a flush is triggered either by your policy, by another policy, or manually.
- `start()`: Executes when the flush policy is enabled and added to the client. This is a good place to start background operations, make async calls, and configure things before execution.

The `FlushPolicy` should have a `shouldFlush` boolean value. When this is set to true, the client attempts to upload events. Each policy should reset this value to `false` according to its own logic, although it's pretty common to do it inside the `reset` method.

```dart

> import 'package:analytics/event.dart';
> import 'package:analytics/flush_policies/flush_policy.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/event.dart';
import 'package:segment_analytics/flush_policies/flush_policy.dart';
```

class FlushOnScreenEventsPolicy extends FlushPolicy {

  @override
  onEvent(RawEvent event) {
    // Only flush when a screen even happens
    if (event is ScreenEvent) {
      this.shouldFlush = true;
    }
  }

  @override
  reset() {
    // Superclass will reset the shouldFlush value so that the next screen event triggers a flush again
    // But you can also reset the value whenever, say another event comes in or after a timeout
    super.reset();
  }
}
```

## Custom logging

By default, any logging is done through the standard Flutter logging mechanism. To customize logging, you can build your own logger, which must implement the `LogTarget` mixin. For example:

```dart

> import 'package:analytics/logger.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/logger.dart';
```

  // ...
}

void customWarningLog(String msg) {
  // ...
}

void customErrorLog(String msg) {
  // ...
}

class CustomLogger with LogTarget {
  @override
  void parseLog(LogMessage log) {
    switch (log.kind) {
      case LogFilterKind.debug:
        customDebugLog("Segment: ${log.message}");
        break;
      case LogFilterKind.warning:
        customWarningLog("Segment: ${log.message}");
        break;
      case LogFilterKind.error:
        customErrorLog("Segment: ${log.message}");
        break;
    }
  }
}

// Set the default logger to use the CustomLogger
LogFactory.logger = CustomLogger();
```

## Handling errors

You can handle analytics client errors through the `errorHandler` option.

The error handler configuration receives a function which gets called whenever an error happens on the analytics client. It receives an Exception, that extends one of the errors from [errors.dart](packages/core/lib/errors.dart).

You can use this error handling to trigger different behaviors in the client when a problem occurs. For example if the client gets rate limited, you could use the error handler to swap flush policies to be less aggressive.

```dart

> import 'package:analytics/errors.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/errors.dart';
```

//...

final flushPolicies = [CountFlushPolicy(5), TimerFlushPolicy(500)];

void errorHandler(Exception error) {
  if (error is NetworkServerLimited) {
    // Remove all flush policies
    analytics.removeFlushPolicy(analytics.getFlushPolicies());
    // Add less persistent flush policies
    analytics.addFlushPolicy([
      CountFlushPolicy(100),
      TimerFlushPolicy(5000)
    ]);
  }
}

final analytics = createClient(Configuration(writeKey),
  errorHandler: errorHandler,
  flushPolicies: flushPolicies);
```

### Reporting errors from plugins

Plugins can also report errors to the handler by using the [`.error`](https://github.com/segmentio/analytics_flutter/blob/main/packages/core/lib/analytics.dart#L52){:target="_blank"} function of the analytics client. Segment recommends you to use the `PluginError` for consistency, and to attach the `innerError` with the actual exception that was hit.

```dart

> import 'package:analytics/errors.dart';

Reviewer 2: ```suggestion
import 'package:segment_analytics/errors.dart';
```

//...

try {
  distinctId = await mixpanel.getDistinctId();
} catch (e) {
  analytics.error(
    PluginError('Error: Mixpanel error calling getDistinctId', e)
  );
}
```

## Platfom specific info

### Web

`analytics_flutter` on web checks for `Analytics.JS` userInfo cookies/localStorage and reuses the `anonymousId` data.

LocalStorage recovery only works when running in the same domain/subdomain.

## Example app

See the [example app](https://github.com/segmentio/analytics_flutter/blob/main/example/README.md){:target="_blank"} to check a full test app of how to integrate Analytics-Flutter into your own Flutter app.

> 

Reviewer 2: It might be worth adding a section here for people upgrading from pilot as there's a few steps they have to do. Not sure if it should be at the end or as part of the install instructions. Here's what I wrote in the readme:

```suggestion
## Upgrading from Pilot

	Since pilot phase we have renamed the package of this library from `analytics` to `segment_analytics`. Some changes have to be applied after upgrading to the v1 package:
	
	In your `pubspec.yaml` remove the `analytics` package and use `segment_analytics` instead.
	
	```diff
	-   analytics:
	-     git:
	-       url: https://github.com/segmentio/analytics_flutter
	-       ref: main
	-       path: packages/core
	+   segment_analytics: ^1.0.1
	```
	
	In your dart files change the imports from `package:analytics` to `package:segment_analytics`:
	
	```diff
	- import 'package:analytics/client.dart';
	+ import 'package:segment_analytics/client.dart';
	```
```

---
title: Analytics-Flutter
hidden: false
---

> info ""
> The Analytics-Flutter library is currently in public beta and is governed by Segment's [First Access and Beta Preview Terms](https://www.twilio.com/en-us/legal/tos){:target="_blank"}. For more information, see the [Analytics-Flutter GitHub repository](https://github.com/segmentio/analytics_flutter){:target="_blank"}.

> warning "Updated package for pilot users"
> If you've been using Analytics-Flutter since the pilot phase, see [Upgrading from pilot](#upgrading-from-pilot) to use the updated version of Analytics-Flutter. 

> Analytics-Flutter enables you to seamlessly add Segment analytics to your Flutter app.

Reviewer 3: ```suggestion
Analytics-Flutter lets you add Segment analytics to your Flutter app.
```
This is subjective, but I'd ditch the "seamlessly," it's giving projection of ease.

        kind: logKind);
    return event;
  }
}
```

As it overrides the `execute` method, this `Logger` calls `log` for every event going through the timeline.

### Supported plugins 
  
You can use these plugins to meet your tracking needs:
  
| Plugin      | Package     |
| ----------- | ----------- |
| [Adjust](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_adjust){:target="_blank"}      | `analytics_plugin_adjust`|
| [AppsFlyer](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_appsflyer){:target="_blank"}    | `analytics_plugin_appsflyer`|
| [Firebase](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_firebase){:target="_blank"}      | `analytics_plugin_firebase`|
| [IDFA](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_idfa){:target="_blank"}     | `analytics_plugin_idfa` |
| [Android Advertising ID](https://github.com/segmentio/analytics_flutter/tree/master/packages/plugins/plugin_advertising_id){:target="_blank"} | `analytics_plugin_advertising-id` |
  
  
## Controlling upload with flush policies

> You can use `FlushPolicies` to more granurily control when events upload.

Reviewer 3: ```suggestion
You can use `FlushPolicies` to more granularly control when events upload.
```

This means only the first policy to reach `shouldFlush` gets to trigger a flush at a time. In the example above, when either the event count gets to 5 or the timer reaches 500ms, whatever comes first triggers a flush.

Segment has several standard FlushPolicies:
- `CountFlushPolicy` triggers whenever a certain number of events is reached.
- `TimerFlushPolicy` triggers on an interval of milliseconds.
- `StartupFlushPolicy` triggers on client startup only.

## Adding or removing policies

One of the main advantanges of `FlushPolicies` is that you can add and remove policies whenever you want. This is powerful when you want to reduce or increase the amount of flushes. 

For example, you might want to disable flushes if you detect the user has no network:

```dart
if (isConnected) {
  analytics.addFlushPolicy(policiesIfNetworkIsUp);
} else {
  analytics.removeFlushPolicy(policiesIfNetworkIsUp)
}
```

### Creating your own flush policies

You can create a custom `FlushPolicy` for your application needs by implementing the  `FlushPolicy` interface. You can also extend the `FlushPolicyBase` class that already creates and handles the `shouldFlush` value reset.

> A `FlushPolicy` only needs to implement 1 method:

Reviewer 3: ```suggestion
A `FlushPolicy` only needs to implement one method:
```

- `onEvent(RawEvent event)`: Gets called on every event tracked by your client.

A `FlushPolicy` can optionally implement:
- `reset()`: Calls after a flush is triggered either by your policy, by another policy, or manually.
- `start()`: Executes when the flush policy is enabled and added to the client. This is a good place to start background operations, make async calls, and configure things before execution.

> The `FlushPolicy` should have a `shouldFlush` boolean value. When this is set to true, the client attempts to upload events. Each policy should reset this value to `false` according to its own logic, although it's pretty common to do it inside the `reset` method.

Reviewer 3: ```suggestion
The `FlushPolicy` should have a `shouldFlush` boolean value. When this is set to `true`, the client attempts to upload events. Each policy should reset this value to `false` according to its own logic, although it's pretty common to do it inside the `reset` method.
```

    // But you can also reset the value whenever, say another event comes in or after a timeout
    super.reset();
  }
}
```

## Custom logging

By default, any logging is done through the standard Flutter logging mechanism. To customize logging, you can build your own logger, which must implement the `LogTarget` mixin. For example:

```dart
import 'package:segment_analytics/logger.dart';

void customDebugLog(String msg) {
  // ...
}

void customWarningLog(String msg) {
  // ...
}

void customErrorLog(String msg) {
  // ...
}

class CustomLogger with LogTarget {
  @override
  void parseLog(LogMessage log) {
    switch (log.kind) {
      case LogFilterKind.debug:
        customDebugLog("Segment: ${log.message}");
        break;
      case LogFilterKind.warning:
        customWarningLog("Segment: ${log.message}");
        break;
      case LogFilterKind.error:
        customErrorLog("Segment: ${log.message}");
        break;
    }
  }
}

// Set the default logger to use the CustomLogger
LogFactory.logger = CustomLogger();
```

## Handling errors

You can handle analytics client errors through the `errorHandler` option.

> The error handler configuration receives a function which gets called whenever an error happens on the analytics client. It receives an Exception, that extends one of the errors from [errors.dart](https://github.com/segmentio/analytics_flutter/blob/main/packages/core/lib/errors.dart){:target="_blank"}.

Reviewer 3: ```suggestion
The error handler configuration receives a function which gets called whenever an error happens on the analytics client. It receives an Exception that extends one of the errors from [errors.dart](https://github.com/segmentio/analytics_flutter/blob/main/packages/core/lib/errors.dart){:target="_blank"}.
```

> You can use this error handling to trigger different behaviors in the client when a problem occurs. For example if the client gets rate limited, you could use the error handler to swap flush policies to be less aggressive.

Reviewer 3: ```suggestion
You can use this error handling to trigger different behaviors in the client when a problem occurs. For example, if the client gets rate limited, you could use the error handler to swap flush policies to be less aggressive.
```
